packages/passport-interface/src/api/makeRequest.ts

```diff
@@ -0,0 +1,119 @@
+import { getErrorMessage } from "@pcd/util";
+import urlJoin from "url-join";
+import { APIResult, GetResultValue, ResultMapper } from "./apiResult";
+import { POST } from "./constants";
+
+/**
+ * Wrapper of {@link httpRequest} that sends a GET request.
+ */
+export async function httpGet<T extends APIResult<unknown, unknown>>(
+  url: string,
+  opts: ResultMapper<T>,
+  query?: object
+): Promise<T> {
+  return httpRequest<T>(
+    urlJoin(
+      url,
+      "?" + new URLSearchParams((query as Record<string, string>) ?? {})
+    ),
+    opts
+  );
+}
+
+/**
+ * Wrapper of {@link httpRequest} that sends a POST request.
+ */
+export async function httpPost<T extends APIResult<unknown, unknown>>(
+  url: string,
+  opts: ResultMapper<T>,
+  postBody?: object
+): Promise<T> {
+  return httpRequest<T>(url, opts, postBody);
+}
+
+/**
+ * Shorthand for a {@link httpGet} whose error type is a string.
+ */
+export async function httpGetSimple<TResult>(
+  url: string,
+  onValue: GetResultValue<APIResult<TResult, string>>,
+  query?: object
+): Promise<APIResult<TResult, string>> {
+  return httpGet<APIResult<TResult, string>>(
+    url,
+    {
+      onValue,
+      onError: async (resText) => ({ error: resText, success: false })
+    },
+    query
+  );
+}
+
+/**
+ * Shorthand for a {@link httpPost} whose error type is a string.
+ */
+export async function httpPostSimple<TResult>(
+  url: string,
+  onValue: GetResultValue<APIResult<TResult, string>>,
+  postBody?: object
+): Promise<APIResult<TResult, string>> {
+  return httpPost<APIResult<TResult, string>>(
+    url,
+    {
+      onValue,
+      onError: async (resText) => ({ error: resText, success: false })
+    },
+    postBody
+  );
+}
+
+/**
+ * Sends a non-blocking HTTP request to the given URL, either a POST
+ * or a GET, with the given body, and converts it into a {@link APIResult}.
+ *
+ * Never rejects.
+ */
+async function httpRequest<T extends APIResult<unknown, unknown>>(
+  url: string,
+  opts: ResultMapper<T>,
+  postBody?: object
+): Promise<T> {
+  let requestOptions: RequestInit = {
+    method: "GET"
+  };
+
+  if (postBody != null) {
+    requestOptions = {
+      ...requestOptions,
+      ...POST,
+      body: JSON.stringify(postBody)
+    };
+  }
+
+  try {
+    // @todo - prevent logspam in the same way we prevent it
+    // on the server. otherwise, these show up in test logs.
+    console.log(`[FETCH]`, url, requestOptions.method);
+
+    const res = await fetch(url, requestOptions);
+    const resText = await res.text();
+
+    if (!res.ok) {
+      console.error("error fetching", url, res.status, resText);
+      return await opts.onError(resText, res.status);
+    }
+
+    return await opts.onValue(resText);
+  } catch (e) {
+    console.error("error fetching", url, e);
+
+    // should I do 1 more extra try/catch around this?
+    return await opts.onError(getErrorMessage(e), undefined);
```

Comment:
the web http fetch API is pretty much a black box - an error thrown by it will not contain any useful information in its stack trace, so its error message is enough. the server will never return a stack trace.

personally I find it much simpler to reason about web api requests in the frontend of the application if I don't have to worry about the dozens of ways an api request can fail, and the several ways that failure can be communicated - either with status codes or with thrown errors. that's why I'm encapsulating the result of an API call into this data structure that either contains a result or an error.

Reply:
That's a very valid and practical reason, and I agree.  I think the concerns I've been thinking through aren't more important than simplicity here, particularly for a relatively small codebase and small team on a web app.  The rest of this comment you can ignore if you want, but continue if you're interested in the alternative arguments I found myself having to review and discard:

- Considering the network libraries a black-box is spoken like someone who's never had to fix a buggy network library. :)  I have, but I think the risk of that here is low.
- There is some application code inside of the relevant try/catch which could fail and lead to an error like "null pointer" (or whatever JS uses for that) which benefits from a stack trace.  The amount of such code is pretty small, though and of course the stack trace should be in the log.
- In addition to the log, an important property I had to reconsider here is that this is a dynamic web app.  If you needed more diagnostic info, you could easily push an update to get it.  That's a very different situation than if you were trying to debug a log from a client-side app with a long roll-out cycle, or even worse an SDK embedded inside someone else's client-side app.

I've been in all of those situations before, which has made me twitchy about error handling and good diagnostics.  I'm convinced your pragmatic approach is the right one for this situation, though.

Reply:
I also worry about situations where some higher-level code needs to distinguish between "the network request failed", "the request succeeded but we got a 5xx error", and "the request succeeded and we got a 4xx error", and I don't want to do that by comparing strings.

The subscription/feed request code is one place where we might need to make these distinctions (though right now we don't). I'm confident that this system is flexible enough that if we need to augment the error report with more structured data on a case-by-case basis, we can do it.